### PR TITLE
Added save and load functions to the framework.

### PR DESCRIPTION
**Pull request by MarcCote**
This PR allows you to save and load an experiment. Basically, it allows to save/load the state of the following classes: `Trainer`, `Optimizer`, `BatchScheduler`, `Loss`, `Model, and`Task`.

``` python
trainer = ... # Create a trainer
trainer.train()  # Train it for some time
trainer.save(path)  # Manually save the trainer (or via a task)
# Crash/Quit
trainer = ... # Rebuild the trainer
trainer.load(path)  # Load previous state of the trainer
trainer.train()  # Resume training
```

For a complete example, check function `test_resume_experiment` in `smartlearner/tests/test_smartlearner.py` .

**Note:** this PR break backward compatibility if your models were defining the `load` method. To be coherent with the rest of the framework `load` is now a method (opposed to a classmethod) that "replaces" the parameters value of an existing instance of `Model`. To create a new instance from data saved on disk, one can use the new method `create` (to be implemented by the subclasses).

### CODE REVIEW

**Comment by MarcCote** [comment]
@ASalvail this one

File: smartlearner/batch_schedulers.py [hunk 1]
```diff
@@ -47,6 +48,18 @@ def __iter__(self):
             self.shared_batch_count.set_value(batch_count)
             yield batch_count + 1
 
+    def save(self, savedir="./"):
+        state = {"version": 1,
```
Comment (ASalvail) [inline smartlearner/batch_schedulers.py:52]: Could it be interesting to hash the class code?
Maybe it's too much too soon.

File: smartlearner/interfaces/model.py [hunk 2]
```diff
@@ -30,8 +22,15 @@ def parameters(self):
...
         raise NotImplementedError("Subclass of 'Model' must implement 'save(path)'.")
 
-    @abstractclassmethod
+    @abstractmethod
     def load(self, path):
+        """ Loads model information from disk. """
         raise NotImplementedError("Subclass of 'Model' must implement 'load(path)'.")
+
+    @classmethod
+    def create(cls, path):
```
Comment (ASalvail) [inline smartlearner/interfaces/model.py]: Why is it 1) public; 2) not incorporated into load?

File: smartlearner/interfaces/optimizer.py [hunk 3]
```diff
@@ -79,3 +87,13 @@ def _apply_modifiers(self, list_modifiers, objects_to_modify):
             updates.update(modifier.updates)
 
         return updates
+
+    def save(self, path):
+        # TODO: should we save directions and params modifiers?
```
Comment (ASalvail) [inline smartlearner/interfaces/optimizer.py]: I hate those names so much. I still can't remember what is a param modifiers.
As for your question, I guess we need to save the full internal state. Then again, it depends on when a save is made. If it's inbetween epochs, it might be necessary only for a few optimizer to keep the whole thing.

File: smartlearner/interfaces/view.py [hunk 4]
```diff
@@ -9,11 +9,13 @@ def __init__(self):
         self.updates = OrderedDict()
         self.value = None
         self.last_update = -1
+        self.last_epoch = -1
 
     def view(self, status):
-        if self.last_update != status.current_update:
+        if self.last_update != status.current_update or self.last_epoch != status.current_epoch:
```
Comment (ASalvail) [inline smartlearner/interfaces/view.py:15]: Aren't `update` and `epochs` updated at the same time?

File: smartlearner/optimizers/adagrad.py [hunk 5]
```diff
@@ -54,3 +58,16 @@ def _get_directions(self):
             directions[param] = (self.lr/root_sum_squared) * direction
 
         return directions
+
+    def _save(self, path):
+        state = {"version": 1}
+        for k, param in self.parameters.items():
+            state[k] = param.get_value()
+
+        np.savez(pjoin(path, 'adagrad.npz'), **state)
```
Comment (ASalvail) [inline smartlearner/optimizers/adagrad.py]: `_save` should return a `state` dictionary that `save` will handle. No point calling `np.savez` all the time with maybe different conventions. `self.__name__` might be useful?

File: smartlearner/optimizers/adagrad.py [hunk 6]
```diff
@@ -54,3 +58,16 @@ def _get_directions(self):
...
+
+    def _load(self, path):
+        state = np.load(pjoin(path, 'adagrad.npz'))
```
Comment (ASalvail) [inline smartlearner/optimizers/adagrad.py]: Same for `_load` which should take a state dictionary as input.

File: smartlearner/tasks.py [hunk 7]
```diff
@@ -23,6 +25,15 @@ def init(self, status):
     def finished(self, status):
         print("Training done in {:.03f} sec.".format(time() - self.start_time))
 
+    def save(self, path):
+        state = {"version": 1,
+                 "start_time": self.start_time}
+        utils.save_dict_to_json_file(pjoin(path, "print_training_duration.json"), state)
```
Comment (ASalvail) [inline smartlearner/tasks.py:31]: Why is this one using JSON while the other are using `np.savez`?

File: smartlearner/trainer.py [hunk 8]
```diff
@@ -42,6 +45,26 @@ def build_theano_graph(self):
                                       name="learn")
         #theano.printing.pydotprint(self._learn, '{0}_learn_{1}'.format(self._optimizer.loss.model.__class__.__name__, theano.config.device), with_ids=True)
 
+    def save(self, path):
+        savedir = utils.create_folder(pjoin(path, "training"))
+        self.status.save(savedir)
+        self._optimizer.save(savedir)
+        self._batch_scheduler.save(savedir)
+
```
Comment (ASalvail) [inline smartlearner/trainer.py:53]: Isn't the model missing?

File: smartlearner/utils.py [hunk 9]
```diff
@@ -17,14 +18,41 @@ def sharedX(value, name=None, borrow=True, keep_on_cpu=False):
...
+
+
+def json_numpy_obj_hook(dct):
+    if isinstance(dct, dict) and '__ndarray__' in dct:
+        return np.array(dct['__ndarray__'])
+
+    return dct
+
+
 def save_dict_to_json_file(path, dictionary):
```
Comment (ASalvail) [inline smartlearner/utils.py:36]: Why yet another way to save to file that isn't used out of the tests?

**Comment by ASalvail** [comment]
There you go, some questions/suggestions/insights.

File: smartlearner/batch_schedulers.py (hunk 1 above)
Comment (MarcCote) [inline smartlearner/batch_schedulers.py:52]: I found version better suited because on load we can do some check and adapt the code accordingly (e.g. version 1 was missing some attribute 'xyz' so we should add it whenever loading code from version <2.

But keeping the hash of the code is an interesting idea even though the hash will change as soon as we make small modification (e.g. spacing). I'll write a Issue so we can keep it in mind but for now I think manual version is more manageable.

File: smartlearner/interfaces/model.py (hunk 2 above)
Comment (MarcCote) [inline smartlearner/interfaces/model.py]: The class method `create` first creates a new instance of the model using the hyperparameters found in `path` (probably a `.json` file), then it sets the parameters of this new instance (via the `load` function) according to the parameters saved in `path`(probably a `.npy` file). Finally, it returns the new instance.

Versus the `load` method that simply sets the parameters that have been saved in `path`.

When resuming a training, we first construct the 'trainer's graph' (i.e. we instantiate the model, the batch_scheduler, the optimizer, the trainer and all the tasks.), then we called `trainer.load` to bring back the training state (i.e. replaces parameters, if needed, of the model, the optimizer, the batch_scheduler, the trainer and the task).

File: smartlearner/interfaces/optimizer.py (hunk 3 above)
Comment (MarcCote) [inline smartlearner/interfaces/optimizer.py]: I think the use case we had for param modifiers is clipping a param (e.g. needs to be > 0). I know you can do it with a directions modifier but you'll have to make sure it is the last modifier to be executed before the params update.

Regarding the question, we'll let the subclasses of `Optimizer` decide if they need to save them.

File: smartlearner/interfaces/view.py (hunk 4 above)
Comment (MarcCote) [inline smartlearner/interfaces/view.py:15]: No `current_update` and `current_epoch` are not updated at the same time. (See the [trainer's loop](https://github.com/SMART-Lab/smartlearner/blob/master/smartlearner/trainer.py#L62)).

The only time that would matter is if we have a trainer that doesn't have any update :P. i think we should keep it like this, better safe than sorry. Also, since it is a `or` statement and it performs a short-circuit evaluation, it is as fast.

File: smartlearner/optimizers/adagrad.py (hunk 5 above)
Comment (MarcCote) [inline smartlearner/optimizers/adagrad.py]: Good idea.

Comment (MarcCote) [inline smartlearner/optimizers/adagrad.py]: @ASalvail: Actually, should we name it `getstate` than?

File: smartlearner/trainer.py (hunk 8 above)
Comment (MarcCote) [inline smartlearner/trainer.py:53]: The trainer doesn't have an reference to the model. The model is saved via the optimizer.

File: smartlearner/tasks.py (hunk 7 above)
Comment (MarcCote) [inline smartlearner/tasks.py:31]: Actually we should used JSON everywhere, especially since I've added a `NumpyEncoder` and `json_numpy_obj_hook` in `smartlearner/utils.py`.

One reason sometime I don't is weights matrices filled up the JSON and is not human readable anymore. So I tend to save the hyperparams in JSON and parameters like the weights of the model and ADAGRAD's parameters to .npy.

What do you think?

File: smartlearner/utils.py (hunk 9 above)
Comment (MarcCote) [inline smartlearner/utils.py:36]: Those are implicitly used when calling `save_dict_to_json_file` and `load_dict_from_json_file` with a dictionary containing any numpy array.

File: smartlearner/interfaces/model.py (hunk 2 above)
Comment (ASalvail) [inline smartlearner/interfaces/model.py]: > When resuming a training, we first construct the 'trainer's graph' (i.e. we instantiate the model, the batch_scheduler, the optimizer, the trainer and all the tasks.), then we called trainer.load to bring back the training state (i.e. replaces parameters, if needed, of the model, the optimizer, the batch_scheduler, the trainer and the task).

So we don't need `create`?

File: smartlearner/optimizers/adagrad.py (hunk 5 above)
Comment (ASalvail) [inline smartlearner/optimizers/adagrad.py]: Sure, why not. It might be useful for other stuff.

File: smartlearner/tasks.py (hunk 7 above)
Comment (ASalvail) [inline smartlearner/tasks.py:31]: That I don't really care for human readability, so long as the titles of the files gives enough info for you to know what's actually inside. If you're scared to forget what was your hyperparameters, you could always code a small exploration function that would load a file in a dict to be able to have a look at it inside ipython.

Conclusion: choose one, I don't really care which one.

File: smartlearner/utils.py (hunk 9 above)
Comment (ASalvail) [inline smartlearner/utils.py:36]: Which are never used in the code. That's my point.

File: smartlearner/interfaces/model.py (hunk 2 above)
Comment (MarcCote) [inline smartlearner/interfaces/model.py]: What about if you want to use a trained model afterwards? The create method is useful.

File: smartlearner/utils.py (hunk 9 above)
Comment (MarcCote) [inline smartlearner/utils.py:36]: Yes, they are used. For instance: https://github.com/SMART-Lab/smartlearner/pull/62/files#diff-8ebe7f074fcf17a46f8e635294cf4782R31 and https://github.com/SMART-Lab/smartlearner/pull/62/files#diff-8ebe7f074fcf17a46f8e635294cf4782R34

**Comment by MarcCote** [comment]
@ppoulin91 I added the following [test](https://github.com/SMART-Lab/smartlearner/pull/62/files#diff-774bda69065f6264bd7df61f72f2a475R82) to check your comment "[Define sigma_t, decaying at each timestep](https://github.com/SMART-Lab/smartlearner/pull/62/files#diff-21cc657d898c077b79e74d5b0e322588R48)" but it's failing. can you have a look?

**Comment by ppoulin91** [comment]
@MarcCote I'll have a look

File: smartlearner/interfaces/batch_scheduler.py [hunk 10]
```diff
@@ -25,3 +22,11 @@ def givens(self):
     @abstractmethod
     def __iter__(self):
         raise NotImplementedError("Subclass of 'BatchScheduler' must implement '__iter__()'.")
+
+    @abstractmethod
+    def save(self, savedir="./"):
+        raise NotImplementedError("Subclass of 'BatchScheduler' must implement 'save(savedir)'.")
```
Comment (ASalvail) [inline smartlearner/interfaces/batch_scheduler.py]: I'd leave those optionnal. I get that ideally this is what should be done, but I'm not sure we should enforce it.

File: smartlearner/interfaces/direction_modifier.py [hunk 11]
```diff
@@ -20,6 +20,14 @@ def _get_updates(self):
     def apply(self, directions):
         raise NotImplementedError("Subclass of 'DirectionModifier' must implement 'apply(directions)'.")
 
+    @abstractmethod
+    def save(self, path):
+        raise NotImplementedError("Subclass of 'DirectionModifier' must implement 'save(path)'.")
```
Comment (ASalvail) [inline smartlearner/interfaces/direction_modifier.py]: Same here, would leave those optional.
